File: 1627-graph-connectivity-with-threshold/1627-graph-connectivity-with-threshold.js
/**
 * @param {number} n
 * @param {number} threshold
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var areConnected = function(n, threshold, queries) {
    const parent = Array(n + 1).fill(-1).map((_, i)=> i)
    const find = (x)=> {
        if(parent[x] !== x)
            parent[x] = find(parent[x])
        return parent[x]
    }
    const union = (a, b) => {
        const pa = find(a)
        const pb = find(b)
        if(pa !== pb)
            parent[pa] = pb
    }
    for(let i = threshold + 1; i <= n; i++){
        for(let j = i * 2; j <= n; j+=i)
            union(i, j)
    }
    
    return queries.map(([x, y]) => find(x) === find(y))
};

File: 1627-graph-connectivity-with-threshold/1627-graph-connectivity-with-threshold.py
class Solution:
    def areConnected(self, n: int, threshold: int, queries: List[List[int]]) -> List[bool]:
        parent = list(range(n+1))
        def union(x, y):
            p1 = find(x)
            p2 = find(y)
            if p1 != p2:
                parent[p1] = p2
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        
        for x in range(threshold+1, n+1):
            for y in range(x+x, n+1, x):
                union(x, y)
        return [find(x) == find(y) for x, y in queries]